### PR TITLE
Added publish all and publish descendants button

### PR DESCRIPTION
**Pull request by kjuej**

### CODE REVIEW

File: client/store/modules/course.js [hunk 1]
```diff
@@ -48,6 +48,13 @@ getter(function activity() {
   return activities[this.state.activity] || {};
 });
 
+getter(function outlineElements() {
+  const { activities } = this.rootGetters;
+  const structure = this.getters['course/structure'];
+  const type = map(structure, 'type');
```
Comment (underscope) [inline client/store/modules/course.js]: It should be plural (types or outlineTypes)

File: client/store/modules/course.js [hunk 2]
```diff
@@ -48,6 +48,13 @@ getter(function activity() {
   return activities[this.state.activity] || {};
 });
 
+getter(function outlineElements() {
```
Comment (underscope) [inline client/store/modules/course.js]: outlineActivities ;)

File: client/components/course/Sidebar/Body.vue [hunk 3]
```diff
@@ -70,9 +114,33 @@ export default {
...
+      Promise.each(activities, activity => {
+        this.publishStatus = `Publishing ${activity.data.name}`;
+        return (this.publish(activity));
+      }).then(() => {
+        this.publishing = false;
+        this.publishStatus = '';
+      });
+    },
+    publishConfirmation(activities, type) {
+      const message = this.getMessage(type);
```
Comment (underscope) [inline client/components/course/Sidebar/Body.vue]: No need to resolve message like this (based on type). You can do it by checking activities length compared to outline activities length.

File: client/components/course/Sidebar/Body.vue [hunk 4]
```diff
@@ -70,9 +114,33 @@ export default {
...
+      appChannel.emit('showConfirmationModal', {
+        type: 'publish',
+        message: message,
+        action: () => {
+          this.publishData(activities);
+        }
+      });
+    },
+    getMessage(type) {
+      if (type === 'publishAll') {
```
Comment (underscope) [inline client/components/course/Sidebar/Body.vue]: Ternary expression that checks passed activities length vs outline activities length.

File: client/components/course/Sidebar/Body.vue [hunk 5]
```diff
@@ -70,9 +114,33 @@ export default {
...
-      this.publish(this.activity).then(() => (this.publishing = false));
+      Promise.each(activities, activity => {
+        this.publishStatus = `Publishing ${activity.data.name}`;
+        return (this.publish(activity));
+      }).then(() => {
+        this.publishing = false;
+        this.publishStatus = '';
+      });
+    },
+    publishConfirmation(activities, type) {
```
Comment (underscope) [inline client/components/course/Sidebar/Body.vue]: Let's always do the confirmation, even if it's one.

File: client/components/course/Sidebar/Body.vue [hunk 6]
```diff
@@ -70,9 +114,33 @@ export default {
...
+      const message = this.getMessage(type);
+      appChannel.emit('showConfirmationModal', {
+        type: 'publish',
+        message: message,
+        action: () => {
+          this.publishData(activities);
+        }
+      });
+    },
+    getMessage(type) {
```
Comment (underscope) [inline client/components/course/Sidebar/Body.vue]: What message?

File: client/components/course/Sidebar/Body.vue (hunk 5 above)
Comment (underscope) [inline client/components/course/Sidebar/Body.vue]: No need to extract it into separate method.

File: client/components/course/Sidebar/Body.vue [hunk 7]
```diff
@@ -36,32 +63,49 @@
...
       return this.getMetadata(this.activity);
     },
-    publishStatus() {
+    publishedAt() {
       let { publishedAt } = this.activity;
       return publishedAt
         ? `Published on ${fecha.format(new Date(publishedAt), 'M/D/YY HH:mm')}`
         : 'Not published';
+    },
+    activityWithDescendants() {
```
Comment (underscope) [inline client/components/course/Sidebar/Body.vue]: You can convert this to one liner...

File: client/components/course/Settings/index.vue [hunk 8]
```diff
@@ -50,25 +59,33 @@
 <script>
 import api from '../../../api/course';
 import CloneModal from './CloneModal';
+import CircularProgress from 'components/common/CircularProgress';
 import EventBus from 'EventBus';
 import General from './General';
 import JSZip from 'jszip';
 import { mapActions, mapGetters } from 'vuex-module';
 import saveAs from 'save-as';
+import publishConfirmation from 'components/common/mixins/publish';
```
Comment (ibebic) [inline client/components/course/Settings/index.vue]: Order

File: client/components/course/Settings/index.vue [hunk 9]
```diff
@@ -50,25 +59,33 @@
 <script>
 import api from '../../../api/course';
 import CloneModal from './CloneModal';
+import CircularProgress from 'components/common/CircularProgress';
```
Comment (ibebic) [inline client/components/course/Settings/index.vue]: Order

File: client/components/course/Settings/index.vue [hunk 10]
```diff
@@ -50,25 +59,33 @@
 <script>
 import api from '../../../api/course';
```
Comment (ibebic) [inline client/components/course/Settings/index.vue]: If you're only using `getContentInventory` from  `api` component, you can destructure it.

File: client/components/course/Sidebar/Body.vue [hunk 11]
```diff
@@ -3,14 +3,41 @@
...
+        <ul class="dropdown-menu">
+          <li>
+            <a
+              @click="publishConfirmation(activityWithDescendants)"
+              href="#">
+              Publish descendants
+            </a>
+          </li>
+          <li>
+            <a
```
Comment (ibebic) [inline client/components/course/Sidebar/Body.vue]: This can fit in one line

File: client/components/course/Sidebar/Body.vue [hunk 12]
```diff
@@ -3,14 +3,41 @@
...
+        </a>
+        <a
+          :disabled="publishing"
+          class="btn btn-primary dropdown-toggle"
+          data-toggle="dropdown">
+          <span class="caret"></span>
+        </a>
+        <ul class="dropdown-menu">
+          <li>
+            <a
```
Comment (ibebic) [inline client/components/course/Sidebar/Body.vue]: This can fit in one line

File: client/components/course/Sidebar/Body.vue [hunk 13]
```diff
@@ -37,42 +64,50 @@
 
 <script>
 import { mapActions, mapGetters } from 'vuex-module';
+import { getDescendants } from 'utils/activity';
 import CircularProgress from 'components/common/CircularProgress';
 import Discussion from './Discussion';
 import fecha from 'fecha';
 import Meta from 'components/common/Meta';
 import Relationship from './Relationship';
+import publishConfirmation from 'components/common/mixins/publish';
```
Comment (ibebic) [inline client/components/course/Sidebar/Body.vue]: Order

File: client/components/course/Sidebar/Body.vue [hunk 14]
```diff
@@ -37,42 +64,50 @@
...
       return this.getMetadata(this.activity);
     },
-    publishStatus() {
+    publishedAt() {
       let { publishedAt } = this.activity;
       return publishedAt
         ? `Published on ${fecha.format(new Date(publishedAt), 'M/D/YY HH:mm')}`
         : 'Not published';
+    },
+    activityWithDescendants({ activity } = this) {
```
Comment (ibebic) [inline client/components/course/Sidebar/Body.vue]: Can you put outlineActivities here as well?

File: client/components/common/mixins/publish.vue [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+    publishConfirmation(activities = [this.activity]) {
+      const message = this.getPublishMessage(activities);
+      appChannel.emit('showConfirmationModal', {
+        type: 'publish',
+        message: message,
+        action: () => {
+          this.publishing = true;
+          Promise.each(activities, activity => {
+            this.publishStatus = `Publishing ${activity.data.name}`;
+            return (this.publish(activity));
```
Comment (ibebic) [inline client/components/common/mixins/publish.vue]: Are parenthesis around `this.publish(activity)` necessary?

File: client/components/course/Sidebar/Body.vue [hunk 16]
```diff
@@ -2,15 +2,38 @@
   <div class="body">
     <div class="publish-container">
       <div class="publish-date">
-        <circular-progress v-if="publishing"></circular-progress>
-        <span v-else>{{ publishStatus }}</span>
+        <circular-progress v-if="isPublishing"></circular-progress>
+        <span v-else>{{ publishedAtMessage }}</span>
+      </div>
+      <div class="btn-group">
```
Comment (underscope) [inline client/components/course/Sidebar/Body.vue]: Extract publish btn to separate component

File: client/components/course/Settings/index.vue [hunk 17]
```diff
@@ -49,26 +58,32 @@
...
+import CircularProgress from 'components/common/CircularProgress';
 import CloneModal from './CloneModal';
 import EventBus from 'EventBus';
 import General from './General';
 import JSZip from 'jszip';
 import { mapActions, mapGetters } from 'vuex-module';
 import saveAs from 'save-as';
+import confirmPublishing from 'components/common/mixins/publish';
```
Comment (underscope) [inline client/components/course/Settings/index.vue]: This is  in general publishing mixin not just confirmation. Please reflect that with naming.

File: client/components/common/ConfirmationModal.vue [hunk 18]
```diff
@@ -42,8 +45,11 @@ export default {
     };
   },
   computed: {
-    info() {
-      return this.context.item.name;
+    publish() {
```
Comment (underscope) [inline client/components/common/ConfirmationModal.vue]: There shouldn't be a notion of publishing within general confirmation modal. Please make sure it stays generic.

File: client/components/common/mixins/publish.vue [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+      const message = this.getPublishMessage(activities.length);
+      appChannel.emit('showConfirmationModal', {
+        title: 'Publish',
+        message: message,
+        action: () => this.publish(activities)
+      });
+    },
+    publish(activities) {
+      this.isPublishing = true;
+      Promise.each(activities, activity => {
```
Comment (underscope) [inline client/components/common/mixins/publish.vue]: return in case somebody wants to chain in the future

**Pull request closed without merging**